x/ccv/consumer/keeper/throttle_retry.go

```diff
@@ -0,0 +1,109 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+
+	consumertypes "github.com/cosmos/interchain-security/v3/x/ccv/consumer/types"
+)
+
+//
+// Throttling with retries follows a FSM design:
+//
+// 1. "No slash": If no slash record exists, the consumer is permitted to send packets from the pending packets queue.
+// The consumer starts in this state from genesis.
+//
+// 2. On the event that a slash packet is obtained from the head of the pending packets queue and sent,
+// a consumer transitions from "No Slash" to "Standby". A slash record is created upon entry to this state,
+// and the consumer is now restricted from sending anymore packets.
+//
+// The slash packet remains at the head of the pending packets queue within the "Standby" state.
+//
+// - If the consumer receives a V1Result ack from the provider, the consumer checks for a slash record,
+// and if found, the consumer transitions from "Standby" to "No Slash". The slash record is cleared upon this transition,
+// and the slash packet is popped from the pending packets queue.
+//
+// - Else if the consumer receives a reply from the provider that the slash packet was successfully handled,
+// the consumer transitions from "Standby" to "No Slash". The slash record is cleared upon this transition,
+// and the slash packet is popped from the pending packets queue.
+//
+// - Else if the consumer receives a reply from the provider that the slash packet was bounced (not handled),
+// then SlashRecord.WaitingOnReply is set false, and the consumer retries sending the slash packet after a delay period.
+//
+// Once a retry is sent, the consumer enters a new cycle of the "Standby" state and the process repeats.
+//
+// Once the slash packet is successfully handled, the consumer transitions from "Standby" to "No Slash",
+// the slash record is cleared upon this transition, and the slash packet is popped from the pending packets queue.
+//
+// This design is implemented below, and in relay.go under SendPackets() and OnAcknowledgementPacket().
+//
+
+// PacketSendingPermitted returns whether the consumer is allowed to send packets
+// from the pending packets queue.
+func (k Keeper) PacketSendingPermitted(ctx sdktypes.Context) bool {
+	record, found := k.GetSlashRecord(ctx)
+	if !found {
+		// no slash record exists, send is permitted
+		return true
+	}
+	if record.WaitingOnReply {
+		// We are waiting on a reply from provider, block sending
+		return false
+	}
+	// Retry delay period could be implemented as a param, but 1 hour is reasonable
+	retryDelayPeriod := time.Hour
+	timeSinceSend := ctx.BlockTime().Sub(record.SendTime)
+	// If retry delay period has elapsed, we can send again
```

Comment:
I guess by param, you mean a general chain param (e.g., like one of [those](https://www.mintscan.io/cosmos/parameters))?

Would it still make sense to define this as a `const` in this file and then have `throttle_retry_test.go` use this constant? Because, if someone were to change this code now to have `retryDelayPeriod := 3 * time.Hour`, then the following test would fail:
```

	// Elapse retry delay period
	ctx = ctx.WithBlockTime(ctx.BlockTime().Add(2 * time.Hour))

	// Now packet sending is permitted again
	require.True(t, consumerKeeper.PacketSendingPermitted(ctx))
```

Reply:
Yea generally if anyone uses the term "param" in a vague way, they're probably talking about the cosmos sdk type of general chain params you've linked 👍 

As far as defining a constant, seems reasonable to me. Note that I doubt this value will ever need to be changed. And if it's intended to be changed, it should be a proper "param". 

https://github.com/cosmos/interchain-security/commit/03daa2a2277c3a7f333f4e18102456dd2f1a2a3d